### PR TITLE
Add efficient logits_to_keep masking to ZambaForCausalLM for constrained generation

### PR DESCRIPTION
**Pull request by Aniketsy**
#40984 

This PR updates the `ZambaForCausalLM` forward method to efficiently mask logits using the `logits_to_keep` argument. It supports both top-k and index-based selection, setting all other logits to `-inf` after projection.

Please let me know if my approach or fix needs any improvements . I’m open to feedback and happy to make changes based on suggestions.
Thankyou !

### CODE REVIEW

**Comment by Aniketsy** [comment]
@Rocketknight1 Could you please review these changes and let me know if any improvements are needed?

**Comment by Rocketknight1** [comment]
Hi @Aniketsy, I think this is a different issue to #40984! That PR was about not computing logits for certain tokens by skipping the LM head entirely. This PR masks certain logits for a single token, making it more like a `LogitsWarper` that is used in generation.

**Comment by Aniketsy** [comment]
@Rocketknight1 Thanks for clarifying! I’ll look more into this, and I’d really appreciate any further guidance or pointers you can share.
